aya/src/symbols.rs

```diff
@@ -0,0 +1,688 @@
+//! Utilities for resolving symbols from addresses obtained via stack traces
+use std::{
+    cell::RefCell,
+    collections::BTreeMap,
+    fs::File,
+    io::{self, BufRead, BufReader},
+    path::{Path, PathBuf},
+};
+
+use addr2line::{
+    gimli::{EndianRcSlice, RunTimeEndian},
+    Context,
+};
+use lru_cache::LruCache;
+use thiserror::Error;
+
+use crate::util::kernel_symbols;
+
+/// The error typey returned when creating various [`SymbolResolver`]s.
+#[derive(Debug, Error)]
+pub enum CreateResolverError {
+    /// Failed to create a kernel symbol resolver
+    #[error("Error creating KernelSymbolResolver: {error}")]
+    KernelSymbolResolverBuildError {
+        #[source]
+        /// The original io::Error
+        error: io::Error,
+    },
+}
+
+/// Stores all information that was resolved for a specific symbol
+#[derive(Debug, Clone)]
+pub struct SymbolInfo {
```

Comment:
I'm thinking, shouldn't we have different types for kernel and user symbols? Eg `process_id`, `object_address` and `object_path` don't really apply to kernel symbols. Is there an advantage to having the same type?

Reply:
It's mostly there to have a common resolver interface. Also if someone wants to represent a full call stack covering kernel and application side, they could use `Vec<SymbolInfo>` or `Stack<SymbolInfo>` instead of having a hetereogenous list. 

But sure, seperate types would also work.

Reply:
I don't think that the ebpf API can return complete user + stack traces. One would have to do some work to reassemble them, and at that point they could also convert to a unified interface.